program/src/processor/harvest_holder_rewards.rs

```diff
@@ -21,11 +36,231 @@ use crate::instruction::accounts::{Context, HarvestHolderRewardsAccounts};
+        "mint"
+    );
+
+    drop(vault_data);
+
+    // stake authority
+    // - must be a signer
+    // - must match the authority on the stake account
+
+    require!(
+        ctx.accounts.stake_authority.is_signer,
+        ProgramError::MissingRequiredSignature,
+        "stake authority",
+    );
+
+    require!(
+        ctx.accounts.stake_authority.key == &stake.authority,
+        StakeError::InvalidAuthority,
+        "stake authority",
+    );
+
+    // vault authority
+    // - derivation must match
+
+    let signer_bump = [config.vault_authority_bump];
+    let vault_signer = create_vault_pda(ctx.accounts.config.key, &signer_bump, program_id)?;
+
+    require!(
+        ctx.accounts.vault_authority.key == &vault_signer,
+        StakeError::InvalidAuthority,
+        "vault authority",
+    );
+
+    // holder rewards (for the vault token account)
+    // - owner must be the rewards program
+    // - derivation must match
+
+    require!(
+        ctx.accounts.holder_rewards.owner == &paladin_rewards_program_client::ID,
+        ProgramError::InvalidAccountOwner,
+        "holder rewards",
+    );
+
+    let (derivation, _) = Pubkey::find_program_address(
+        &["holder".as_bytes(), ctx.accounts.vault.key.as_ref()],
+        &paladin_rewards_program_client::ID,
+    );
+
+    require!(
+        ctx.accounts.holder_rewards.key == &derivation,
+        ProgramError::InvalidSeeds,
+        "holder rewards",
+    );
+
+    // Determine the holder rewards.
+
+    let holder_rewards = HolderRewards::try_from(ctx.accounts.holder_rewards)?;
+    let rewards = calculate_eligible_rewards(
+        holder_rewards.last_accumulated_rewards_per_token,
+        stake.last_seen_holder_rewards_per_token(),
+        stake.amount,
+    )?;
+
+    // Withdraw the holder rewards to the destination account.
+
+    if rewards != 0 {
+        // update the last seen holder rewards
+        stake.set_last_seen_holder_rewards_per_token(
+            holder_rewards.last_accumulated_rewards_per_token,
+        );
+
+        // Rewards are stored on the `vault` token account. We need to first withdraw the excess lamports
+        // from the vault token account to the vault authority account. Then transfer the rewards amount
+        // to the destination account and send the remaining excess lamports back to the vault token
+        // account.
+
+        let withdraw_ix = withdraw_excess_lamports(
+            ctx.accounts.token_program.key,
+            ctx.accounts.vault.key,
+            ctx.accounts.vault_authority.key,
+            ctx.accounts.vault_authority.key,
+            &[],
+        )?;
+
+        let signer_seeds = get_vault_pda_signer_seeds(ctx.accounts.config.key, &signer_bump);
+        // Stores the current lamports of the vault authority account before the withdraw so
+        // we can calculate the amount of rewards withdrawn.
+        let current_lamports = ctx.accounts.vault_authority.lamports();
```

Comment:
nit: any chance we can make this a bit more clear? It becomes hard to follow after a few steps below.

```suggestion
        let vault_authority_starting_lamports = ctx.accounts.vault_authority.lamports();
```